### PR TITLE
[SPARK-27402][SQL][test-hadoop3.2][test-maven]  Fix hadoop-3.2 test issue(except the hive-thriftserver module)

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

This pr fix hadoop-3.2 test issues(except the `hive-thriftserver` module):
1. Add `hive.metastore.schema.verification` and `datanucleus.schema.autoCreateAll` to HiveConf.
2. hadoop-3.2 support access the Hive metastore from 0.12 to 2.2

After [SPARK-27176](https://issues.apache.org/jira/browse/SPARK-27176) and this PR, we upgraded the built-in Hive to 2.3 when enabling the Hadoop 3.2+ profile. This upgrade fixes the following issues:
- [HIVE-6727](https://issues.apache.org/jira/browse/HIVE-6727): Table level stats for external tables are set incorrectly.
- [HIVE-15653](https://issues.apache.org/jira/browse/HIVE-15653): Some ALTER TABLE commands drop table stats.
- [SPARK-12014](https://issues.apache.org/jira/browse/SPARK-12014): Spark SQL query containing semicolon is broken in Beeline.
- [SPARK-25193](https://issues.apache.org/jira/browse/SPARK-25193): insert overwrite doesn't throw exception when drop old data fails.
- [SPARK-25919](https://issues.apache.org/jira/browse/SPARK-25919): Date value corrupts when tables are "ParquetHiveSerDe" formatted and target table is Partitioned.
- [SPARK-26332](https://issues.apache.org/jira/browse/SPARK-26332): Spark sql write orc table on viewFS throws exception.
- [SPARK-26437](https://issues.apache.org/jira/browse/SPARK-26437): Decimal data becomes bigint to query, unable to query.

## How was this patch tested?
This pr test Spark’s Hadoop 3.2 profile on jenkins and #24591 test Spark’s Hadoop 2.7 profile on jenkins

This PR close #24591

### CODE REVIEW

File: dev/sparktestsupport/modules.py [hunk 1]
```diff
@@ -72,7 +78,10 @@ def __init__(self, name, dependencies, source_file_regexes, build_profile_flags=
         self.dependent_modules = set()
         for dep in dependencies:
             dep.dependent_modules.add(self)
-        all_modules.append(self)
+        if name == "hive-thriftserver" and hadoop_version == "hadoop3.2":
```
Comment (wangyum) [inline dev/sparktestsupport/modules.py]: Used to skip hive-thriftserver module for hadoop-3.2.  Will revert this change once we can merge.

File: pom.xml [hunk 2]
```diff
@@ -2819,7 +2819,7 @@
         <hive.version>${hive23.version}</hive.version>
         <hive.version.short>2.3.4</hive.version.short>
         <hive.parquet.group>org.apache.parquet</hive.parquet.group>
-        <hive.parquet.version>1.8.1</hive.parquet.version>
+        <hive.parquet.version>${parquet.version}</hive.parquet.version>
```
Comment (wangyum) [inline pom.xml]: Will revert this change once shade Parquet to sql/core.

**Comment by wangyum** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 3]
```diff
@@ -189,13 +190,15 @@ private[hive] class IsolatedClientLoader(
...
     name.startsWith("javax.sql.") ||
+    name.startsWith("org.apache.derby.") ||
     sharedPrefixes.exists(name.startsWith)
   }
 
   /** True if `name` refers to a spark class that must see specific version of Hive. */
   protected def isBarrierClass(name: String): Boolean =
     name.startsWith(classOf[HiveClientImpl].getName) ||
     name.startsWith(classOf[Shim].getName) ||
+    name.startsWith(classOf[ShimLoader].getName) ||
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Add `org.apache.hadoop.hive.shims.ShimLoader` to BarrierClass, otherwise hadoop-3.2 can't access the Hive metastore from 0.12 to 2.2:
```java
build/sbt "hive/testOnly *.VersionsSuite" -Phadoop-3.2 -Phive
...
[info] - 0.12: create client *** FAILED *** (36 seconds, 207 milliseconds)
[info]   java.lang.reflect.InvocationTargetException:
[info]   at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[info]   at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[info]   at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[info]   at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
[info]   at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:295)
[info]   at org.apache.spark.sql.hive.client.HiveClientBuilder$.buildClient(HiveClientBuilder.scala:58)
[info]   at org.apache.spark.sql.hive.client.VersionsSuite.$anonfun$new$7(VersionsSuite.scala:130)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:105)
[info]   at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
[info]   at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[info]   at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
[info]   at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
[info]   at org.scalatest.FunSuite.runTest(FunSuite.scala:1560)
[info]   at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[info]   at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
[info]   at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
[info]   at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
[info]   at org.scalatest.Suite.run(Suite.scala:1147)
[info]   at org.scalatest.Suite.run$(Suite.scala:1129)
[info]   at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
[info]   at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[info]   at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
[info]   at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:54)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:54)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:507)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:296)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:286)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:748)
[info]   Cause: java.lang.RuntimeException: org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.RuntimeException: java.lang.IllegalArgumentException: Unrecognized Hadoop major version number: 3.2.0
[info]   at org.apache.hadoop.hive.ql.session.SessionState.start(SessionState.java:286)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.newState(HiveClientImpl.scala:187)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.<init>(HiveClientImpl.scala:119)
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala [hunk 4]
```diff
@@ -27,7 +27,8 @@ package object client {
   // scalastyle:off
   private[hive] object hive {
     case object v12 extends HiveVersion("0.12.0")
-    case object v13 extends HiveVersion("0.13.1")
+    case object v13 extends HiveVersion("0.13.1",
+      exclusions = Seq("commons-httpclient:commons-httpclient"))
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala]: Workaround: [NOT FOUND  ] commons-httpclient#commons-httpclient;3.0.1!commons-httpclient.jar (1ms)

https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104649/testReport/org.apache.spark.sql.hive/HiveExternalCatalogVersionsSuite/backward_compatibility/
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104649/testReport/org.apache.spark.sql.hive.execution/Hive_2_1_DDLSuite/SPARK_21617__ALTER_TABLE_for_non_compatible_DataSource_tables/

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 5]
```diff
@@ -189,13 +190,15 @@ private[hive] class IsolatedClientLoader(
     (name.startsWith("com.google") && !name.startsWith("com.google.cloud")) ||
     name.startsWith("java.") ||
     name.startsWith("javax.sql.") ||
+    name.startsWith("org.apache.derby.") ||
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: I think this shouldn't be added without a strong reason as discussed at https://github.com/apache/spark/pull/20944#issuecomment-481962005 if I am not mistaken.

Comment (HyukjinKwon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Can we try to find the cause and backport it within Hive side because it looks a bug in Hive 2.3.4? We can put some inputs into Hive side to make 2.3.5. It's mush smaller job for Hive side instead of dealing with HIVE-16391.

Comment (HyukjinKwon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: If there's some delays during ^, we can discuss to consider this as a temporary bandaid fix but I am worried if we happen to introduce some issues hard to debug (like I met in Livy).

**Comment by wangyum** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala (hunk 5 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: We could fix it at Hive side: https://issues.apache.org/jira/browse/HIVE-21639

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala (hunk 4 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala]: It's our jenkins issue, will rever this change.

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by gatorsmile** [comment]
ping @shaneknapp Could you please help take a look at it?

**Comment by shaneknapp** [comment]
i wiped the ivy/maven cache on this box and this problem should go away.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 6]
```diff
@@ -868,14 +873,21 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
         // ALTER TABLE SET/UNSET TBLPROPERTIES invalidates some Hive specific statistics, but not
         // Spark specific statistics. This is triggered by the Hive alterTable API.
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: update the comments here

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 7]
```diff
@@ -868,14 +873,21 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
         val numRows = extractStatsPropValues(describeResult, "numRows")
-        assert(numRows.isDefined && numRows.get == -1, "numRows is lost")
-        val rawDataSize = extractStatsPropValues(describeResult, "rawDataSize")
-        assert(rawDataSize.isDefined && rawDataSize.get == -1, "rawDataSize is lost")
-
-        if (analyzedBySpark) {
+        if (HiveUtils.isHive23) {
+          assert(numRows.isDefined && numRows.get == 500)
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Finally, Hive fixed this annoying bug. The statistics are very important for us too. Please add this in the migration guide and document the behavior difference when they used different Hadoop profile.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 8]
```diff
@@ -100,8 +100,13 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
             .asInstanceOf[HiveTableRelation]
 
           val properties = relation.tableMeta.ignoredProperties
-          assert(properties("totalSize").toLong <= 0, "external table totalSize must be <= 0")
-          assert(properties("rawDataSize").toLong <= 0, "external table rawDataSize must be <= 0")
+          if (HiveUtils.isHive23) {
+            assert(properties("totalSize").toLong > 0, "external table totalSize must be > 0")
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: what is the number? How does Hive collect this info?

**Comment by gatorsmile** [comment]
My review is not finished. I will review it this weekend. If I do not leave the comments, please ping me using IMs

File: dev/sparktestsupport/modules.py [hunk 9]
```diff
@@ -18,9 +18,15 @@
...
 import re
+import os
 
 all_modules = []
 
+if os.environ.get("AMPLAB_JENKINS"):
+    hadoop_version = os.environ.get("AMPLAB_JENKINS_BUILD_PROFILE", "hadoop2.7")
+else:
+    hadoop_version = os.environ.get("HADOOP_PROFILE", "hadoop2.7")
+
```
Comment (gatorsmile) [inline dev/sparktestsupport/modules.py]: We should have a log message to show which profile we are printing; otherwise, it is hard for us to know which profile is activated.

File: docs/sql-migration-guide-upgrade.md [hunk 10]
```diff
@@ -24,6 +24,22 @@ license: |
 
 ## Upgrading From Spark SQL 2.4 to 3.0
 
+  - Since Spark 3.0, we upgraded the built-in Hive to 2.3 for hadoop-3.x. This upgrade fixes some Hive issues:
```
Comment (gatorsmile) [inline docs/sql-migration-guide-upgrade.md]: >  for hadoop-3.x
when enabling the Hadoop 3.2+ profile.

File: dev/sparktestsupport/modules.py (hunk 9 above)
Comment (wangyum) [reply dev/sparktestsupport/modules.py]: Yes. we already have these log message:
image
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105106/consoleFull

Comment (gatorsmile) [reply dev/sparktestsupport/modules.py]: What I mean, we should have a log message to show which profile is being actually used. 

Just relying on the command parameter lists does not sound very reliable

File: docs/sql-migration-guide-upgrade.md [hunk 11]
```diff
@@ -24,6 +24,22 @@ license: |
 
 ## Upgrading From Spark SQL 2.4 to 3.0
 
+  - Since Spark 3.0, we upgraded the built-in Hive to 2.3 when enabling the Hadoop 3.2+ profile. This upgrade fixes some Hive issues:
```
Comment (gatorsmile) [inline docs/sql-migration-guide-upgrade.md]: the following issues:

File: dev/sparktestsupport/modules.py (hunk 9 above)
Comment (wangyum) [reply dev/sparktestsupport/modules.py]: OK. updated to:
image

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 12]
```diff
@@ -1371,7 +1371,7 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
       // if (isUsingHiveMetastore) {
       //  assert(storageFormat.properties.get("path") === expected)
       // }
-      assert(storageFormat.locationUri === Some(expected))
+      assert(Some(storageFormat.locationUri.get.getPath) === Some(expected.getPath))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Do we need to add `Some`?

What is the reason we need to change this test case?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 13]
```diff
@@ -191,7 +191,17 @@ private[hive] class HiveClientImpl(
...
-  def conf: HiveConf = state.getConf
+  def conf: HiveConf = {
+    val hiveConf = state.getConf
+    // Hive changed the default of datanucleus.schema.autoCreateAll from true to false
+    // and hive.metastore.schema.verification from false to true since Hive 2.0.
+    // For details, see the JIRA HIVE-6113, HIVE-12463 and HIVE-1841.
+    // We hard-code these configurations here to allow bin/spark-shell, bin/spark-sql
+    // and sbin/start-thriftserver.sh automatically creates Derby metastore.
+    hiveConf.setBoolean("hive.metastore.schema.verification", false)
+    hiveConf.setBoolean("datanucleus.schema.autoCreateAll", true)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Could you add a test case to ensure that we still respect the values set from users?

For example, if users set a different value, we do not reset them to our values.

**Comment by gatorsmile** [comment]
My review is finished. Overall, it looks good to me. After we ensure that maven can pass all the tests, we also need to rerun SBT as well.

**Comment by gatorsmile** [comment]
@wangyum Thank you for your hard work!!!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 12 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: image
Add `Some` to support both Hive 1.2 and 2.3.

File: dev/sparktestsupport/modules.py (hunk 1 above)
Comment (srowen) [reply dev/sparktestsupport/modules.py]: Maybe leave a TODO here just to try to make sure that doesn't get lost

File: docs/sql-migration-guide-upgrade.md [hunk 14]
```diff
@@ -24,6 +24,22 @@ license: |
 
 ## Upgrading From Spark SQL 2.4 to 3.0
 
+  - Since Spark 3.0, we upgraded the built-in Hive to 2.3 when enabling the Hadoop 3.2+ profile. This upgrade fixes the following issues:
```
Comment (srowen) [inline docs/sql-migration-guide-upgrade.md]: Does this need to go in the migration guide? it's helpful info but does it help a user know what they need to do differently?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 12 above)
Comment (srowen) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: `storageFormat.locationUri.map(_.getPath)` then

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcFilterSuite.scala [hunk 15]
```diff
@@ -76,15 +77,27 @@ class HiveOrcFilterSuite extends OrcTest with TestHiveSingleton {
...
   }
 
-  private def checkFilterPredicate
+  private def checkFilterPredicateWithDiffHiveVersion
       (predicate: Predicate, stringExpr: String)
       (implicit df: DataFrame): Unit = {
     def checkLogicalOperator(filter: SearchArgument) = {
-      assert(filter.toString == stringExpr)
+      if (HiveUtils.isHive23) {
+        assert(filter.toString == stringExpr.replace("\n", ", "))
```
Comment (srowen) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcFilterSuite.scala]: Would this actually also pass on Hive 1? meaning would the replace just do nothing? if so this wouldn't need to be special case, but I don't know what the test output is like

Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcFilterSuite.scala]: No. It wont pass on Hive 1.

File: dev/sparktestsupport/modules.py (hunk 1 above)
Comment (wangyum) [reply dev/sparktestsupport/modules.py]: Add to release note?

File: docs/sql-migration-guide-upgrade.md (hunk 14 above)
Comment (wangyum) [reply docs/sql-migration-guide-upgrade.md]: Add to release note?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 13 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I update it to:
```scala
  def conf: HiveConf = {
    val hiveConf = state.getConf
    // Hive changed the default of datanucleus.schema.autoCreateAll from true to false
    // and hive.metastore.schema.verification from false to true since Hive 2.0.
    // For details, see the JIRA HIVE-6113, HIVE-12463 and HIVE-1841.
    // For the production environment. Either isDefaultMSUri or isDerbyMS should not be true.
    // We hard-code hive.metastore.schema.verification and datanucleus.schema.autoCreateAll to allow
    // bin/spark-shell, bin/spark-sql and sbin/start-thriftserver.sh to automatically create the
    // Derby Metastore when running Spark in the non-production environment.
    val isDefaultMSUri = hiveConf.getVar(METASTOREURIS).equals(METASTOREURIS.defaultStrVal)
    val isDerbyMS = hiveConf.getVar(METASTORECONNECTURLKEY).startsWith("jdbc:derby")
    if (isDefaultMSUri && isDerbyMS) {
      hiveConf.setBoolean("hive.metastore.schema.verification", false)
      hiveConf.setBoolean("datanucleus.schema.autoCreateAll", true)
    }
    hiveConf
  }
```
Because we don't know it's default value or set by user. so just set it for non-production environment.

File: docs/sql-migration-guide-upgrade.md (hunk 14 above)
Comment (gatorsmile) [reply docs/sql-migration-guide-upgrade.md]: The JIRA lists can be moved to the PR description, but we should document the actual behavior difference in migration guide.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 16]
```diff
@@ -191,7 +191,27 @@ private[hive] class HiveClientImpl(
   }
 
   /** Returns the configuration for the current session. */
-  def conf: HiveConf = state.getConf
+  def conf: HiveConf = {
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: To avoid a regression, could you change it to ?

```Scala
if (!isHive23) {
    state.getConf
  } else {
    ...
  }
```

File: docs/sql-migration-guide-upgrade.md (hunk 14 above)
Comment (wangyum) [reply docs/sql-migration-guide-upgrade.md]: OK. Create [SPARK-27686](https://issues.apache.org/jira/browse/SPARK-27686) to update migration guide

**Comment by wangyum** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master.

File: dev/sparktestsupport/modules.py [hunk 17]
```diff
@@ -15,9 +15,17 @@
 # limitations under the License.
 #
 
+from __future__ import print_function
 from functools import total_ordering
 import itertools
 import re
+import os
+
+if os.environ.get("AMPLAB_JENKINS"):
```
Comment (HyukjinKwon) [inline dev/sparktestsupport/modules.py]: @wangyum, this will shows the info every time this modules is imported. why did we do this here?

Comment (HyukjinKwon) [reply dev/sparktestsupport/modules.py]: okay. it's a temp fix so I'm fine. I will make a followup to handle https://github.com/apache/spark/pull/24639/files

Comment (wangyum) [reply dev/sparktestsupport/modules.py]: Skip the `hive-thriftserver` module when running the hadoop-3.2 test. Will remove it in another PR: https://github.com/apache/spark/pull/24628/files

Comment (wangyum) [reply dev/sparktestsupport/modules.py]: Thank you @HyukjinKwon
